homeassistant/auth/__init__.py

```diff
@@ -253,19 +337,44 @@ def async_create_access_token(self,
         """Create a login flow."""
         auth_provider = self._providers[handler]
 
-        return await auth_provider.async_credential_flow(context)
+        return await auth_provider.async_login_flow(context)
 
     async def _async_finish_login_flow(
-            self, context: Optional[Dict], result: Dict[str, Any]) \
-            -> Optional[models.Credentials]:
-        """Result of a credential login flow."""
+            self, flow: LoginFlow, result: Dict[str, Any]) \
+            -> Dict[str, Any]:
+        """Return a user as result of login flow."""
```

Comment:
We need to add a lot more documentation to this method . What is happening, when and why.

Reply:
So I am slightly confused why we are even making changes for this method?

All auth provider login flows will create `async_start_mfa(flow_result)`. The MFA flow will show more forms if necessary and eventually will pass the original flow result on if MFA has passed. That means that this method will not see any new data?

Reply:
I changed design, now the auth provider (not the base class) will not know about mfa, they just call `async_finish` after validate credential. The `_async_finish_login_flow` will do the heave lift, if need mfa then change the result type to form by calling `async_step_select_mfa_module`